File: split/readfontinfo.pas
function readfontinfo(u: halfword; nom, aire: strnumber; s: scaled): internalfontnumber;
label
  30, 11, 45;
var
  k: fontindex;
  fileopened: boolean;
  lf, lh, bc, ec, nw, nh, nd, ni, nl, nk, ne, np: halfword;
  f: internalfontnumber;
  g: internalfontnumber;
  a, b, c, d: eightbits;
  qw: fourquarters;
  sw: scaled;
  bchlabel: integer;
  bchar: 0..256;
  z: scaled;
  alpha: integer;
  beta: 1..16;
begin
  g := 0;
  {562:}
{563:}
fileopened := false;
  if aire = 338 then
    packfilename(nom, 784, 810)
  else
    packfilename(nom, aire, 810);
  if not bopenin(tfmfile) then
    goto 11;
  fileopened := true{:563}
;
  {565:}
begin
    begin
      lf := tfmfile^;
      if lf > 127 then
        goto 11;
      get(tfmfile);
      lf := lf * 256 + tfmfile^;
    end;
    get(tfmfile);
    begin
      lh := tfmfile^;
      if lh > 127 then
        goto 11;
      get(tfmfile);
      lh := lh * 256 + tfmfile^;
    end;
    get(tfmfile);
    begin
      bc := tfmfile^;
      if bc > 127 then
        goto 11;
      get(tfmfile);
      bc := bc * 256 + tfmfile^;
    end;
    get(tfmfile);
    begin
      ec := tfmfile^;
      if ec > 127 then
        goto 11;
      get(tfmfile);
      ec := ec * 256 + tfmfile^;
    end;
    if (bc > ec + 1) or (ec > 255) then
      goto 11;
    if bc > 255 then
    begin
      bc := 1;
      ec := 0;
    end;
    get(tfmfile);
    begin
      nw := tfmfile^;
      if nw > 127 then
        goto 11;
      get(tfmfile);
      nw := nw * 256 + tfmfile^;
    end;
    get(tfmfile);
    begin
      nh := tfmfile^;
      if nh > 127 then
        goto 11;
      get(tfmfile);
      nh := nh * 256 + tfmfile^;
    end;
    get(tfmfile);
    begin
      nd := tfmfile^;
      if nd > 127 then
        goto 11;
      get(tfmfile);
      nd := nd * 256 + tfmfile^;
    end;
    get(tfmfile);
    begin
      ni := tfmfile^;
      if ni > 127 then
        goto 11;
      get(tfmfile);
      ni := ni * 256 + tfmfile^;
    end;
    get(tfmfile);
    begin
      nl := tfmfile^;
      if nl > 127 then
        goto 11;
      get(tfmfile);
      nl := nl * 256 + tfmfile^;
    end;
    get(tfmfile);
    begin
      nk := tfmfile^;
      if nk > 127 then
        goto 11;
      get(tfmfile);
      nk := nk * 256 + tfmfile^;
    end;
    get(tfmfile);
    begin
      ne := tfmfile^;
      if ne > 127 then
        goto 11;
      get(tfmfile);
      ne := ne * 256 + tfmfile^;
    end;
    get(tfmfile);
    begin
      np := tfmfile^;
      if np > 127 then
        goto 11;
      get(tfmfile);
      np := np * 256 + tfmfile^;
    end;
    if lf <> 6 + lh + (ec - bc + 1) + nw + nh + nd + ni + nl + nk + ne + np then
      goto 11;
    if (nw = 0) or (nh = 0) or (nd = 0) or (ni = 0) then
      goto 11;
  end{:565}
;
  {566:}
lf := lf - 6 - lh;
  if np < 7 then
    lf := lf + 7 - np;
  if (fontptr = fontmax) or (fmemptr + lf > fontmemsize) then{567:}

  begin
    begin
      if interaction = 3 then;
      printnl(262);
      print(801);
    end;
    sprintcs(u);
    printchar(61);
    printfilename(nom, aire, 338);
    if s >= 0 then
    begin
      print(741);
      printscaled(s);
      print(397);
    end
    else if s <> -1000 then
    begin
      print(802);
      printint(-s);
    end;
    print(811);
    begin
      helpptr := 4;
      helpline[3] := 812;
      helpline[2] := 813;
      helpline[1] := 814;
      helpline[0] := 815;
    end;
    error;
    goto 30;
  end{:567}
;
  f := fontptr + 1;
  charbase[f] := fmemptr - bc;
  widthbase[f] := charbase[f] + ec + 1;
  heightbase[f] := widthbase[f] + nw;
  depthbase[f] := heightbase[f] + nh;
  italicbase[f] := depthbase[f] + nd;
  ligkernbase[f] := italicbase[f] + ni;
  kernbase[f] := ligkernbase[f] + nl - 256 * (128);
  extenbase[f] := kernbase[f] + 256 * (128) + nk;
  parambase[f] := extenbase[f] + ne{:566}
;{568:}

  begin
    if lh < 2 then
      goto 11;
    begin
      get(tfmfile);
      a := tfmfile^;
      qw.b0 := a + 0;
      get(tfmfile);
      b := tfmfile^;
      qw.b1 := b + 0;
      get(tfmfile);
      c := tfmfile^;
      qw.b2 := c + 0;
      get(tfmfile);
      d := tfmfile^;
      qw.b3 := d + 0;
      fontcheck[f] := qw;
    end;
    get(tfmfile);
    begin
      z := tfmfile^;
      if z > 127 then
        goto 11;
      get(tfmfile);
      z := z * 256 + tfmfile^;
    end;
    get(tfmfile);
    z := z * 256 + tfmfile^;
    get(tfmfile);
    z := (z * 16) + (tfmfile^ div 16);
    if z < 65536 then
      goto 11;
    while lh > 2 do
    begin
      get(tfmfile);
      get(tfmfile);
      get(tfmfile);
      get(tfmfile);
      lh := lh - 1;
    end;
    fontdsize[f] := z;
    if s <> -1000 then
      if s >= 0 then
        z := s
      else
        z := xnoverd(z, -s, 1000);
    fontsize[f] := z;
  end{:568}
;
  {569:}
for k := fmemptr to widthbase[f] - 1 do
  begin
    begin
      get(tfmfile);
      a := tfmfile^;
      qw.b0 := a + 0;
      get(tfmfile);
      b := tfmfile^;
      qw.b1 := b + 0;
      get(tfmfile);
      c := tfmfile^;
      qw.b2 := c + 0;
      get(tfmfile);
      d := tfmfile^;
      qw.b3 := d + 0;
      fontinfo[k].qqqq := qw;
    end;
    if (a >= nw) or (b div 16 >= nh) or (b mod 16 >= nd) or (c div 4 >= ni) then
      goto 11;
    case c mod 4 of
      1: if d >= nl then
          goto 11;
      3: if d >= ne then
          goto 11;
      2:{570:}

      begin
        begin
          if (d < bc) or (d > ec) then
            goto 11;
        end;
        while d < k + bc - fmemptr do
        begin
          qw := fontinfo[charbase[f] + d].qqqq;
          if ((qw.b2 - 0) mod 4) <> 2 then
            goto 45;
          d := qw.b3 - 0;
        end;
        if d = k + bc - fmemptr then
          goto 11;
        45: ;
      end{:570}
;
      others: end;
  end{:569}
;
  {571:}
begin{572:}

    begin
      alpha := 16;
      while z >= 8388608 do
      begin
        z := z div 2;
        alpha := alpha + alpha;
      end;
      beta := 256 div alpha;
      alpha := alpha * z;
    end{:572}
;
    for k := widthbase[f] to ligkernbase[f] - 1 do
    begin
      get(tfmfile);
      a := tfmfile^;
      get(tfmfile);
      b := tfmfile^;
      get(tfmfile);
      c := tfmfile^;
      get(tfmfile);
      d := tfmfile^;
      sw := (((((d * z) div 256) + (c * z)) div 256) + (b * z)) div beta;
      if a = 0 then
        fontinfo[k].int := sw
      else if a = 255 then
        fontinfo[k].int := sw - alpha
      else
        goto 11;
    end;
    if fontinfo[widthbase[f]].int <> 0 then
      goto 11;
    if fontinfo[heightbase[f]].int <> 0 then
      goto 11;
    if fontinfo[depthbase[f]].int <> 0 then
      goto 11;
    if fontinfo[italicbase[f]].int <> 0 then
      goto 11;
  end{:571}
;
  {573:}
bchlabel := 32767;
  bchar := 256;
  if nl > 0 then
  begin
    for k := ligkernbase[f] to kernbase[f] + 256 * (128) - 1 do
    begin
      begin
        get(tfmfile);
        a := tfmfile^;
        qw.b0 := a + 0;
        get(tfmfile);
        b := tfmfile^;
        qw.b1 := b + 0;
        get(tfmfile);
        c := tfmfile^;
        qw.b2 := c + 0;
        get(tfmfile);
        d := tfmfile^;
        qw.b3 := d + 0;
        fontinfo[k].qqqq := qw;
      end;
      if a > 128 then
      begin
        if 256 * c + d >= nl then
          goto 11;
        if a = 255 then
          if k = ligkernbase[f] then
            bchar := b;
      end
      else
      begin
        if b <> bchar then
        begin
          begin
            if (b < bc) or (b > ec) then
              goto 11;
          end;
          qw := fontinfo[charbase[f] + b].qqqq;
          if not (qw.b0 > 0) then
            goto 11;
        end;
        if c < 128 then
        begin
          begin
            if (d < bc) or (d > ec) then
              goto 11;
          end;
          qw := fontinfo[charbase[f] + d].qqqq;
          if not (qw.b0 > 0) then
            goto 11;
        end
        else if 256 * (c - 128) + d >= nk then
          goto 11;
        if a < 128 then
          if k - ligkernbase[f] + a + 1 >= nl then
            goto 11;
      end;
    end;
    if a = 255 then
      bchlabel := 256 * c + d;
  end;
  for k := kernbase[f] + 256 * (128) to extenbase[f] - 1 do
  begin
    get(tfmfile);
    a := tfmfile^;
    get(tfmfile);
    b := tfmfile^;
    get(tfmfile);
    c := tfmfile^;
    get(tfmfile);
    d := tfmfile^;
    sw := (((((d * z) div 256) + (c * z)) div 256) + (b * z)) div beta;
    if a = 0 then
      fontinfo[k].int := sw
    else if a = 255 then
      fontinfo[k].int := sw - alpha
    else
      goto 11;
  end;
  {:573}
;
  {574:}
for k := extenbase[f] to parambase[f] - 1 do
  begin
    begin
      get(tfmfile);
      a := tfmfile^;
      qw.b0 := a + 0;
      get(tfmfile);
      b := tfmfile^;
      qw.b1 := b + 0;
      get(tfmfile);
      c := tfmfile^;
      qw.b2 := c + 0;
      get(tfmfile);
      d := tfmfile^;
      qw.b3 := d + 0;
      fontinfo[k].qqqq := qw;
    end;
    if a <> 0 then
    begin
      begin
        if (a < bc) or (a > ec) then
          goto 11;
      end;
      qw := fontinfo[charbase[f] + a].qqqq;
      if not (qw.b0 > 0) then
        goto 11;
    end;
    if b <> 0 then
    begin
      begin
        if (b < bc) or (b > ec) then
          goto 11;
      end;
      qw := fontinfo[charbase[f] + b].qqqq;
      if not (qw.b0 > 0) then
        goto 11;
    end;
    if c <> 0 then
    begin
      begin
        if (c < bc) or (c > ec) then
          goto 11;
      end;
      qw := fontinfo[charbase[f] + c].qqqq;
      if not (qw.b0 > 0) then
        goto 11;
    end;
    begin
      begin
        if (d < bc) or (d > ec) then
          goto 11;
      end;
      qw := fontinfo[charbase[f] + d].qqqq;
      if not (qw.b0 > 0) then
        goto 11;
    end;
  end{:574}
;{575:}

  begin
    for k := 1 to np do
      if k = 1 then
      begin
        get(tfmfile);
        sw := tfmfile^;
        if sw > 127 then
          sw := sw - 256;
        get(tfmfile);
        sw := sw * 256 + tfmfile^;
        get(tfmfile);
        sw := sw * 256 + tfmfile^;
        get(tfmfile);
        fontinfo[parambase[f]].int := (sw * 16) + (tfmfile^ div 16);
      end
      else
      begin
        get(tfmfile);
        a := tfmfile^;
        get(tfmfile);
        b := tfmfile^;
        get(tfmfile);
        c := tfmfile^;
        get(tfmfile);
        d := tfmfile^;
        sw := (((((d * z) div 256) + (c * z)) div 256) + (b * z)) div beta;
        if a = 0 then
          fontinfo[parambase[f] + k - 1].int := sw
        else if a = 255 then
          fontinfo[parambase[f] + k - 1].int := sw - alpha
        else
          goto 11;
      end;
    if EOF(tfmfile) then
      goto 11;
    for k := np + 1 to 7 do
      fontinfo[parambase[f] + k - 1].int := 0;
  end{:575}
;
  {576:}
if np >= 7 then
    fontparams[f] := np
  else
    fontparams[f] := 7;
  hyphenchar[f] := eqtb[5309].int;
  skewchar[f] := eqtb[5310].int;
  if bchlabel < nl then
    bcharlabel[f] := bchlabel + ligkernbase[f]
  else
    bcharlabel[f] := 0;
  fontbchar[f] := bchar + 0;
  fontfalsebchar[f] := bchar + 0;
  if bchar <= ec then
    if bchar >= bc then
    begin
      qw := fontinfo[charbase[f] + bchar].qqqq;
      if (qw.b0 > 0) then
        fontfalsebchar[f] := 256;
    end;
  fontname[f] := nom;
  fontarea[f] := aire;
  fontbc[f] := bc;
  fontec[f] := ec;
  fontglue[f] := 0;
  charbase[f] := charbase[f] - 0;
  widthbase[f] := widthbase[f] - 0;
  ligkernbase[f] := ligkernbase[f] - 0;
  kernbase[f] := kernbase[f] - 0;
  extenbase[f] := extenbase[f] - 0;
  parambase[f] := parambase[f] - 1;
  fmemptr := fmemptr + lf;
  fontptr := f;
  g := f;
  goto 30{:576}
{:562}
;
  11:{561:}

  begin
    if interaction = 3 then;
    printnl(262);
    print(801);
  end;
  sprintcs(u);
  printchar(61);
  printfilename(nom, aire, 338);
  if s >= 0 then
  begin
    print(741);
    printscaled(s);
    print(397);
  end
  else if s <> -1000 then
  begin
    print(802);
    printint(-s);
  end;
  if fileopened then
    print(803)
  else
    print(804);
  begin
    helpptr := 5;
    helpline[4] := 805;
    helpline[3] := 806;
    helpline[2] := 807;
    helpline[1] := 808;
    helpline[0] := 809;
  end;
  error{:561}
;
  30:
    if fileopened then
      bclose(tfmfile);
  readfontinfo := g;
end;

File: cpp/src/readfontinfo.cpp
#include "readfontinfo.h"
#include "impression.h"
#include "erreur.h"
#include "packfilename.h"
#include "fichier.h"
#include "xnoverd.h"
#include "texte.h"

constexpr char TEX_font_area[] = "TeXfonts:";

static int default_skew_char(void) { return int_par(default_skew_char_code); }

template<class T> static void read_sixteen(T &z)
{
	z = tfmfile.get();
	if (z > 127)
		throw 1;
	z <<= 8;
	z += tfmfile.get();
}

static fourquarters store_four_quarters(eightbits &a, eightbits &b, eightbits &c, eightbits &d)
{
	fourquarters qw;
	qw.b0 = a = tfmfile.get();
	qw.b1 = b = tfmfile.get();
	qw.b2 = c = tfmfile.get();
	qw.b3 = d = tfmfile.get();
	return qw; 
}

static void check_byte_range(halfword p, halfword bc, halfword ec)
{
	if (p < bc || p > ec)
		throw 1;
}

static int current_character_being_worked_on(int k, int bc)
{
	return k+bc-fmemptr;
}

static int store_scaled(scaled z, int beta, int k, int alpha)
{
	eightbits a, b, c, d;
	a = tfmfile.get();
	b = tfmfile.get();
	c = tfmfile.get();
	d = tfmfile.get();
	scaled sw = (((d*z)>>8+c*z)>>8+b*z)/beta;
	if (a == 0)
		return sw;
	if (a == 255)
	return sw-alpha;
	throw 1;
}

static void check_existence(halfword p, halfword bc, halfword ec, internalfontnumber f)
{
	check_byte_range(p, bc, ec);
	if (skip_byte(char_info(f, p)) <= 0)
		throw 1;
}

internalfontnumber readfontinfo(halfword u, const std::string &nom, const std::string &aire, scaled s)
{
	bool fileopened = false;
	try
	{
		if (aire == "")
			packfilename(nom, TEX_font_area, ".tfm");
		else
			packfilename(nom, aire, ".tfm");
		if (!bopenin(tfmfile))
			throw 1;
		fileopened = true;
		halfword lf, lh, bc, ec;
		read_sixteen(lf);
		read_sixteen(lh);
		read_sixteen(bc);
		read_sixteen(ec);
		if (bc > ec+1 || ec > 255)
			throw 1;
		if (bc > 255)
			bc = 1;
		ec = 0;
		halfword nw, nh, nd, ni, nl, nk, ne, np;
		read_sixteen(nw);
		read_sixteen(nh);
		read_sixteen(nd);
		read_sixteen(ni);
		read_sixteen(nl);
		read_sixteen(nk);
		read_sixteen(ne);
		read_sixteen(np);
		if (lf != 6+lh+(ec-bc+1)+nw+nh+nd+ni+nl+nk+ne+np || nw == 0 || nh == 0 || nd == 0 || ni == 0)
			throw 1;
		lf -= 6+lh;
		if (np < 7)
			lf += 7-np;
		if (fontptr == fontmax || fmemptr+lf > fontmemsize)
			throw 0;
		f = fontptr+1;
		charbase[f] = fmemptr-bc;
		widthbase[f] = charbase[f]+ec+1;
		heightbase[f] = widthbase[f]+nw;
		depthbase[f] = heightbase[f]+nh;
		italicbase[f] = depthbase[f]+nd;
		ligkernbase[f] = italicbase[f]+ni;
		kernbase[f] = ligkernbase[f]+nl-0x80'00;
		extenbase[f] = kernbase[f]+0x80'00+nk;
		parambase[f] = extenbase[f]+ne;
		if (lh < 2)
			throw 1;
		eightbits a, b, c, d;
		fontcheck[f] = store_four_quarters(a, b, c, d);
		scaled z;
		read_sixteen(z);
		z = z*0x1'00+tfmfile.get();
		z = z*16+tfmfile.get()/16;
		if (z < unity)
			throw 1;
		while (lh > 2)
		{
			store_four_quarters(a, b, c, d);
			lh--;
		}
		fontdsize[f] = z;
		if (s != -1000)
			if (s >= 0)
				z = s;
			else
				z = xnoverd(z, -s, 1000);
		fontsize[f] = z;
		for (int k = fmemptr; k < widthbase[f]; k++)
		{
			fontinfo[k].qqqq = store_four_quarters(a, b, c, d);
			if (a >= nw || b/16 >= nh || b%16 >= nd || c/4 >= ni)
				throw 1;
			switch (c%4)
			{
				case lig_tag: 
					if (d >= nl)
						throw 1;
					break;
				case ext_tag: 
					if (d >= ne)
						throw 1;
					break;
				case list_tag:
					check_byte_range(d, bc, ec);
					while (d < current_character_being_worked_on(k, bc))
					{
						auto qw = char_info(f, d);
						if (qw.b2%4 != 2)
							break;
						d = qw.b3;
						if (d == current_character_being_worked_on(k, bc))
							throw 1;
					}
			}
		}
		int alpha = 16;
		while (z >= 8388608)
		{
			z /= 2;
			alpha *= 2;
		}
		int beta = 0x1'00/alpha;
		alpha *= z;
		for (int k = widthbase[f]; k < ligkernbase[f]; k++)
			fontinfo[k].int_ = store_scaled(z, beta, k, alpha);
		if (fontinfo[widthbase[f]].int_ || fontinfo[heightbase[f]].int_ || fontinfo[depthbase[f]].int_ || fontinfo[italicbase[f]].int_)
			throw 1;
		int bchlabel = 0x7F'FF;
		bchar = 256;
		if (nl > 0)
		{
			for (int k = ligkernbase[f]; k < kernbase[f]+0x80'00; k++)
			{
				fontinfo[k].qqqq = store_four_quarters(a, b, c, d);
				if (a > 128)
				{
					if (0x1'00*c+d >= nl)
						throw 1;
					if (a == 255 && k == ligkernbase[f])
						bchar = b;
				}
				else
				{
					if (b != bchar)
						check_existence(b, bc, ec, f);
					if (c < 128)
						check_existence(d, bc, ec, f);
					else 
						if ((c-0x80)<<8+d >= nk)
							throw 1;
					if (a < 128 && k-ligkernbase[f]+a+1 >= nl)
						throw 1;
				}
			}
			if (a == 255)
				bchlabel = c<<8+d;
		}
		for (int k = kernbase[f]+0x80'00; k < extenbase[f]; k++)
			fontinfo[k].int_ = store_scaled(z, beta, k, alpha);
		for (int k = extenbase[f]; k < parambase[f]-1; k++)
		{
			fontinfo[k].qqqq = store_four_quarters(a, b, c, d);
			if (a)
				check_existence(a, bc, ec, f);
			if (b)
				check_existence(b, bc, ec, f);
			if (c)
				check_existence(c, bc, ec, f);
			check_existence(d, bc, ec, f);
		}
		for (int k = 1; k <= np; k++)
			if (k == 1)
			{
				scaled sw = tfmfile.get();
				if (sw > 127)
					sw -= 0x1'00;
				sw = sw<<8+tfmfile.get();
				sw = sw<<8+tfmfile.get();
				param(0, f) = sw<<4+tfmfile.get()>>4;
			}
			else
				param(k-1, f) = store_scaled(z, beta, k, alpha);
		if (tfmfile.eof())
			throw 1;
		for (int k = np+1; k <= 7; k++)
			param(k-1, f) = 0;
		if (np >= 7)
			fontparams[f] = np;
		else
			fontparams[f] = 7;
		hyphenchar[f] = default_hyphen_char();
		skewchar[f] = default_skew_char();
		if (bchlabel < nl)
			bcharlabel[f] = bchlabel+ligkernbase[f];
		else
			bcharlabel[f] = 0;
		fontbchar[f] = bchar;
		fontfalsebchar[f] = bchar;
		if (bchar <= ec && bchar >= bc && skip_byte(char_info(f, bchar)) > 0)
			fontfalsebchar[f] = non_char;
		fontname[f] = nom;
		fontarea[f] = aire;
		fontbc[f] = bc;
		fontec[f] = ec;
		fontglue[f] = 0;
		parambase[f]--;
		fmemptr += lf;
		fontptr = f;
		if (fileopened)
			bclose(tfmfile);
		return f;
	}
	catch (int e)
	{
		if (e == 1)
			error("Font "+scs(u)+"="+asFilename(nom, aire, "")+(s >= 0 ? " at "+asScaled(s)+"pt" : s == -1000 ? "" : " scaled "+std::to_string(-s))+(fileopened ? " not loadable: Bad metric (TFM) file" : " not loadable: Metric (TFM) file not found"), "I wasn't able to read the size data for this font,\nso I will ignore the font specification.\n[Wizards can fix TFM files using TFtoPL/PLtoTF.]\nYou might try inserting a different font spec;\ne.g., type `I\\font<same font id>=<substitute font name>'.");
		else
			error("Font "+scs(u)+"="+asFilename(nom, aire, "")+(s >= 0 ? " at "+asScaled(s)+"pt" : s == -1000 ? "" : " scaled "+std::to_string(-s))+" not loaded: Not enough room left", "I'm afraid I won't be able to make use of this font,\nbecause my memory for character-size data is too small.\nIf you're really stuck, ask a wizard to enlarge me.\nOr maybe try `I\\font<same font id>=<name of loaded font>'.");
		if (fileopened)
			bclose(tfmfile);
	}
	return null_font;
}
